README.md
# Energy-based models for atomic-resolution protein conformations
Pytorch implementation of [Energy-based models for atomic-resolution protein conformations](https://openreview.net/forum?id=S1e_9xrFvS), accepted to ICLR 2020 (with spotlight). Includes training code, models, datasets, and pre-trained model weights.

![Framework overview](https://dl.fbaipublicfiles.com/protein-ebm/framework_overview.png)

## Dependencies

To install the dependencies for the project, execute
```
pip install -r requirements.txt
```

For reproducibility, we listed the packages used for generating the results in the paper, but other versions of these packages will likely give similar results.

## Downloading the datasets

### 1. Rotamer Library
First, get the [Dunbrack rotamer library](http://dunbrack.fccc.edu/bbdep2010/) (935.0 MB), which is used for negative sampling.
```
wget https://dl.fbaipublicfiles.com/protein-ebm/dunbrak_rotamer.tar.gz
tar xvzf dunbrack_rotamer.tar.gz 
```
If necessary, update `ROTAMER_LIBRARY_PATH` in `config.py` to point to the `original/` directory in the uncompressed `tar.gz` parent directory.

### 2. Protein structures
Next, download the mmCIF files (38.7 GiB) used for training:
```
wget https://dl.fbaipublicfiles.com/protein-ebm/mmcif.tar.gz
tar xvzf mmcif.tar.gz
```
If necessary, update `MMCIF_PATH` in `config.py` to point to the unzipped directory.

### 3. [optional] Pre-trained models
To retreive pre-trained EBM models (48.9 MB) that were evaluated in [our paper](https://openreview.net/forum?id=S1e_9xrFvS), run
```
wget https://dl.fbaipublicfiles.com/protein-ebm/cachedir.tar.gz
tar xvzf cachedir.tar.gz
```

The transformer model evaluated in the paper can then be found at `cachedir/transformer_gmm_uniform/model_130000`. 

## Preprocessing
After downloading the protein structures, please preprocess the data by running
```
python mmcif_utils.py
```
This will generate a set of preprocessed intermediates for each mmCIF file. Additionally, it run a set of unit tests to check correctness of the kinematics operations.

## Training

### Negative Sampling Arguments

EBMs are trained with positive samples and negative samples. The positive samples are the mmCIF files you generated above, but the negative samples are chosen or constructed during training. Our codebase implements several different strategies for negative sampling:
1. uniform: when selected, sample the rotamers uniformly from the Dunbrack rotamer library. If not, sample the rotamers based on their emperical frequencies.
2. weighted\_gauss: when selected, interpolate between candidate rotamers for nearby phi/psi bins by sampling from a Gaussian using weighted means and variances of each nearby bin
3. gmm: when selected, interpolate between candidate rotamers for nearby phi/psi bins by sampling from a Gaussian Mixture Model with mixture weights corresponding to the distances of each bin
4. chi\_mean: when selected, sample candidate rotamers by using mean chi angles from the Dunbrack library and angles 1 standard deviation away

For the experiments reported in [our paper](https://openreview.net/forum?id=S1e_9xrFvS), we interpolated between candidate rotamers using a Guassian Mixture Model, with uniform sampling of the rotamers. This corresponds to the negative sample arguments: `-gmm --uniform`.

### Model Arguments

In [our paper](https://openreview.net/forum?id=S1e_9xrFvS), we report the performance of various model architectures for predicting energies of rotamer configurations. The models we tested include:
1. fc: a MLP architecture
2. s2s: a Set2Set architecture
3. transformer: a transformer architecture
2. graph: a GNN architecture

For our experiments and the released pre-trained model, we use the transformer architecture, which corresponds to `--model transformer`. 

### Training the model
The following script will train a model on a single node:
```
python train.py

## Negative sampling arguments
--neg-sample 1                     # number of negative rotamer samples for rotamer trials (1-1 ratio)
--gmm                              # example negative sampling arugment
--uniform                          # example negative sampling argument

## Model parameters
--encoder-layers 3                 # number of layers

## optimization
--gpus 1                           # number of gpus
--nodes 1                          # number of nodes
--batch-size 196                   # batch size
--end-lr 0.0002                    # learning rate after warmup
--nodes 1

## main parameters
--exp test_ebm                     # experiment name
```

In [our paper](https://openreview.net/forum?id=S1e_9xrFvS), we trained models on 32 different GPUs. Therefore, we've included code in this release for distributed training on multiple nodes. The following script will train a model on 32 GPUs:
```
python train.py

## Negative sampling arguments
--neg-sample 1                     # number of negative rotamer samples for rotamer trials (1-1 ratio)
--gmm                              # example negative sampling arugment
--uniform                          # example negative sampling argument

## Model parameters
--encoder-layers 3                 # number of layers

## optimization
--gpus 1                           # number of gpus
--nodes 4                          # number of nodes
--batch-size 196                   # batch size
--end-lr 0.0002                    # learning rate after warmup
--nodes 1

## distributed training
--node-rank 2                      # the rank of the current node
--master-addr 8.8.8.8              # sets os.environ['MASTER_ADDR']
--master-port 8888                 # sets os.environ['MASTER_PORT']
--slurm                            # if using SLURM

## main parameters
--exp test_ebm                     # experiment name
```

To enable distributed training, set `master-addr` and `master-port` to an accessible machine. Then, run one copy of this script on each node, ensuring that the process has access to all GPUs on that node. Gradients will sync to the `master-addr` and `master-port`.

If you are using SLURM, include the `--slurm` argument. This tells `torch.distributed` to set `init_method='env://'`. Otherwise, `init_method=tcp://localhost:1492` by default.

If you need to modify any of the distributed settings for your infrastructure, see the start of the `main()` and `main_single()` functions in `train.py` where this logic is implemented.


## Visualization
To reproduce the visualizations shown in our paper, run
```
python vis_sandbox.py --exp=transformer_gmm_uniform --resume-iter=130000 --task=[new_model, tsne]
```

where the task can be switched to visualize different results. 

Additional code for creating saliency maps can be found in `scripts/extract_saliency.py` and `scripts/generate_colormap_saliency.py`.


## Quantitative Metrics
To produce the quantiative metrics for performance on the rotamer recovery task (using discrete sampling),
run
```
python vis_sandbox.py --exp=transformer_gmm_uniform --resume-iter=130000 --task=rotamer_trial --sample-mode=rosetta --neg-sample 500 --rotations 10
```
or run
```
python vis_sandbox.py --exp=transformer_gmm_uniform --resume-iter=130000 --task=rotamer_trial --sample-mode=gmm --neg-sample 150  --rotations 10
```
for continuous sampling.

## Citing this work
If you find the EBM useful, please cite [our corresponding paper](https://openreview.net/forum?id=S1e_9xrFvS). Additionally, significant time and effort went into the dataset construction established by previous works. Please cite those papers if you use the associated datasets.

### Our paper (Protein EBM):
```
@inproceedings{
    Du2020Energy-based,
    title={Energy-based models for atomic-resolution protein conformations},
    author={Yilun Du and Joshua Meier and Jerry Ma and Rob Fergus and Alexander Rives},
    booktitle={International Conference on Learning Representations},
    year={2020},
    url={https://openreview.net/forum?id=S1e_9xrFvS}
}
```

### Dunbrack rotamer library:
```
@article{bower1997prediction,
  title={Prediction of protein side-chain rotamers from a backbone-dependent rotamer library: a new homology modeling tool},
  author={Bower, Michael J and Cohen, Fred E and Dunbrack Jr, Roland L},
  journal={Journal of molecular biology},
  volume={267},
  number={5},
  pages={1268--1282},
  year={1997},
  publisher={Elsevier}
}
```

### CullPDB Database
```
@Article{dunbrackresolution,
   Author="G. Wang and R. L. Dunbrack, Jr.",
   Title="PISCES: a protein sequence culling server",
   Journal="Bioinformatics",
   Year="2003",
   Volume="19",
   Pages="1589-1591"
}
```

### Rotamer recovery benchmark
```
@incollection{leaver2013scientific,
  title={Scientific benchmarks for guiding macromolecular energy function improvement},
  author={Leaver-Fay, Andrew and O'Meara, Matthew J and Tyka, Mike and Jacak, Ron and Song, Yifan and Kellogg, Elizabeth H and Thompson, James and Davis, Ian W and Pache, Roland A and Lyskov, Sergey and others},
  booktitle={Methods in enzymology},
  volume={523},
  pages={109--143},
  year={2013},
  publisher={Elsevier}
}
```

## License
protein-ebm is MIT licensed, as found in the LICENSE file in the root directory of this source tree.


CHANGELOG.md
1.0.0 (April 27, 2020)

### Initial release

* Initial release of code for [Energy-based models for atomic-resolution protein conformations](https://openreview.net/forum?id=S1e_9xrFvS), accepted to ICLR 2020 (with spotlight).


train.py
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import argparse
import os
import os.path as osp
import pickle
import random

import numpy as np
from scipy.stats import special_ortho_group

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn.functional as F
from config import MMCIF_PATH
from data import MMCIFTransformer, collate_fn_transformer, collate_fn_transformer_test
from easydict import EasyDict
from mmcif_utils import compute_rotamer_score_planar
from models import RotomerFCModel, RotomerGraphModel, RotomerSet2SetModel, RotomerTransformerModel
from tensorboard import TensorBoardOutputFormat
from tensorflow.python.platform import flags
from torch import nn, optim
from torch.utils.data import DataLoader, Dataset
from tqdm import tqdm
from utils import init_distributed_mode


def add_args(parser):
    #############################
    ##### Training hyperparameters
    #############################

    parser.add_argument(
        "--logdir",
        default="cachedir",
        type=str,
        help="location where log of experiments will be stored",
    )
    parser.add_argument("--exp", default="default", type=str, help="name of experiment")
    parser.add_argument("--resume-iter", default=0, type=int, help="resume value")
    parser.add_argument("--n-epochs", default=10000, type=int, help="number of epochs of training")
    parser.add_argument(
        "--batch-size", default=128, type=int, help="batch size to use during training"
    )
    parser.add_argument("--log-interval", default=50, type=int, help="interval to log results")
    parser.add_argument("--save-interval", default=1000, type=int, help="interval to log results")
    parser.add_argument(
        "--no-train",
        default=False,
        action="store_true",
        help="Instead of training, only test the model",
    )
    parser.add_argument(
        "--no-cuda", default=False, action="store_true", help="do not use GPUs for computations"
    )
    parser.add_argument(
        "--model",
        default="transformer",
        type=str,
        help="model to use during training. options: transformer, fc, s2s (set 2 set) "
        "transformer: transformer model"
        "fc: MLP baseline used in paper"
        "s2s: Set2Set baseline used in paper"
        "graph: GNN baseline used in paper",
    )

    #############################
    ##### Dataset hyperparameters
    #############################

    parser.add_argument("--data-workers", default=4, type=int, help="number of dataloader workers")
    parser.add_argument(
        "--multisample",
        default=16,
        type=int,
        help="number of different rotamers to select" "from an individual protein",
    )

    #############################
    ##### Distributed Training
    #############################

    parser.add_argument("--nodes", default=1, type=int, help="number of nodes for training")

    parser.add_argument("--gpus", default=1, type=int, help="number of gpus per node")
    parser.add_argument("--node-rank", default=0, type=int, help="rank of node")
    parser.add_argument(
        "--master-addr", default="8.8.8.8", type=str, help="address of communicating server"
    )
    parser.add_argument("--port", default=10002, type=int, help="port for communicating server")
    parser.add_argument(
        "--slurm", default=False, action="store_true", help="run experiments on SLURM?"
    )

    #############################
    ##### Transformer hyperparameters
    #############################

    parser.add_argument(
        "--encoder-layers", default=6, type=int, help="number of transformer layers"
    )
    parser.add_argument(
        "--dropout", default=0.0, type=float, help="dropout of attention weights in transformer"
    )
    parser.add_argument(
        "--relu-dropout", default=0.0, type=float, help="chance of dropping out a relu unit"
    )
    parser.add_argument(
        "--no-encoder-normalize-before",
        action="store_true",
        default=False,
        help="do not normalize outputs before the encoder (transformer only)",
    )
    parser.add_argument(
        "--encoder-attention-heads",
        default=8,
        type=int,
        help="number of attention heads (transformer only)",
    )
    parser.add_argument(
        "--attention-dropout", default=0.0, type=float, help="dropout probability for attention"
    )
    parser.add_argument(
        "--encoder-ffn-embed-dim",
        default=1024,
        type=int,
        help="hidden dimension to use in transformer",
    )
    parser.add_argument(
        "--encoder-embed-dim", default=256, type=int, help="original embed dim of element"
    )
    parser.add_argument(
        "--max-size",
        default=64,
        type=str,
        help="number of nearby atoms to attend" "when predicting energy of rotamer",
    )

    #############################
    ##### EBM hyperparameters
    #############################

    parser.add_argument(
        "--neg-sample",
        default=1,
        type=int,
        help="number of negative rotamer samples" " per real data sample (1-1 ratio)",
    )
    parser.add_argument("--l2-norm", default=False, action="store_true", help="norm the energies")
    parser.add_argument(
        "--no-augment",
        default=False,
        action="store_true",
        help="do not augment training data with so3 rotations",
    )

    #############################
    ##### generic model params
    #############################

    parser.add_argument(
        "--start-lr", default=1e-10, type=float, help="initial warmup learning rate"
    )
    parser.add_argument("--end-lr", default=2e-4, type=float, help="end lr of training")
    parser.add_argument(
        "--lr-schedule",
        default="constant",
        type=str,
        help="schedule to anneal the learning rate of transformer."
        " options: constant, inverse_sqrt",
    )
    parser.add_argument("--warmup-itr", default=500, type=int, help="number of warm up iterations")
    parser.add_argument(
        "--single",
        default=False,
        action="store_true",
        help="overfit to a single protein in dataset" "(sanity check on architecture)",
    )
    parser.add_argument(
        "--grad_accumulation", default=1, type=int, help="number of gradient accumulation steps"
    )

    #############################
    ##### Negative sampling
    #############################

    parser.add_argument(
        "--uniform",
        default=False,
        action="store_true",
        help="uniform over all candidate bins in Dunbrak library"
        "as oposed to weighted based off empericial frequency",
    )
    parser.add_argument(
        "--weighted-gauss",
        default=False,
        action="store_true",
        help="given chi and psi angles, iterpolate between nearby bins"
        "based off Gaussian with weighted sum of means/var with weights computed by distance",
    )
    parser.add_argument(
        "--gmm",
        default=False,
        action="store_true",
        help="given chi and psi angles, interpolate between nearby bins"
        "by sampling each nearby bin based off Gaussian with weights computed by distance",
    )
    parser.add_argument(
        "--chi-mean",
        default=False,
        action="store_true",
        help="instead of sampling from Gaussians from bins in the Dunbrak library"
        "just sample the mean of the bins",
    )

    return parser


def average_gradients(model):
    """Averages gradients across workers"""
    size = float(dist.get_world_size())

    for param in model.parameters():
        dist.all_reduce(param.grad.data, op=dist.reduce_op.SUM)
        param.grad.data /= size


def sync_model(model):
    """Sync parameters across models"""
    size = float(dist.get_world_size())

    for param in model.parameters():
        dist.broadcast(param.data, 0)


def compute_lr(it, FLAGS):
    lr = FLAGS.start_lr + min(it / FLAGS.warmup_itr, 1) * (FLAGS.end_lr - FLAGS.start_lr)

    if FLAGS.lr_schedule == "inverse_sqrt":
        if it - FLAGS.warmup_itr > 10:
            lr = lr * ((it - FLAGS.warmup_itr) / 10) ** -0.5
    return lr


def train(
    train_dataloader,
    test_dataloader,
    logger,
    model,
    optimizer,
    FLAGS,
    logdir,
    rank_idx,
    train_structures,
    checkpoint=None,
):
    it = FLAGS.resume_iter
    count_it = 0
    for i in range(FLAGS.n_epochs):

        for node_pos, node_neg in train_dataloader:
            if it % 500 == 499 and ((rank_idx == 0 and FLAGS.gpus > 1) or FLAGS.gpus == 1):
                test_accs, test_losses, test_rotamer = test(
                    test_dataloader, model, FLAGS, logdir, test=False
                )
                kvs = {}
                kvs["test_losses"] = test_losses
                kvs["test_rotamer"] = test_rotamer
                kvs["test_accs"] = test_accs

                string = "Test epoch of  {}".format(i)
                for k, v in kvs.items():
                    string += "{}: {}, ".format(k, v)
                    logger.writekvs(kvs)

                print(string)

            if FLAGS.cuda:
                node_pos = node_pos.cuda()
                node_neg = node_neg.cuda()

            lr = compute_lr(it, FLAGS)

            # update the optimizer learning rate
            for param_group in optimizer.param_groups:
                param_group["lr"] = lr / FLAGS.grad_accumulation

            energy_pos = model.forward(node_pos)
            energy_neg = model.forward(node_neg)
            energy_neg = energy_neg.view(energy_pos.size(0), -1)

            partition_function = -torch.cat([energy_pos, energy_neg], dim=1)

            log_prob = (-energy_pos) - torch.logsumexp(partition_function, dim=1, keepdim=True)
            loss = (-log_prob).mean()
            loss.backward()

            if it % FLAGS.grad_accumulation == 0:
                if FLAGS.gpus > 1:
                    average_gradients(model)

                optimizer.step()
                optimizer.zero_grad()

            it += 1
            count_it += 1

            if it % FLAGS.log_interval == 0 and rank_idx == 0:
                loss = loss.item()
                energy_pos = energy_pos.mean().item()
                energy_neg = energy_neg.mean().item()

                kvs = {}
                kvs["loss"] = loss
                kvs["energy_pos"] = energy_pos
                kvs["energy_neg"] = energy_neg

                string = "Iteration {} with values of ".format(it)

                for k, v in kvs.items():
                    string += "{}: {}, ".format(k, v)
                    logger.writekvs(kvs)

                print(string)

            if it % FLAGS.save_interval == 0 and rank_idx == 0:
                model_path = osp.join(logdir, "model_{}".format(it))
                torch.save(
                    {
                        "model_state_dict": model.state_dict(),
                        "optimizer_state_dict": optimizer.state_dict(),
                        "FLAGS": FLAGS,
                    },
                    model_path,
                )
                print("Saving model in directory....")


def test(test_dataloader, model, FLAGS, logdir, test=False):
    # Load data from a dataloader and then evaluate both the reconstruction MSE and visualize proteins
    cum_corrupt = []
    cum_decorrupt = []
    accs = []
    losses = []
    rotamer_recovery = []
    best_rotamer_recovery = []

    angle_width = np.pi / 180 * 1

    if FLAGS.train:
        neg_sample = 150
    else:
        neg_sample = FLAGS.neg_sample

    if test:
        itr = 1
    else:
        itr = 20

    counter = 0
    for _ in range(itr):
        for node_pos, node_neg, gt_chis, neg_chis, res in tqdm(test_dataloader):
            counter += 1

            if FLAGS.cuda:
                node_pos = node_pos.cuda()
                node_neg = node_neg.cuda()

            with torch.no_grad():
                energy_pos = model.forward(node_pos)
                energy_neg = model.forward(node_neg)
                energy_neg = energy_neg.view(energy_pos.size(0), neg_sample)
                partition_function = -torch.cat([energy_pos, energy_neg], dim=1)
                idx = torch.argmax(partition_function, dim=1)
                sort_idx = torch.argsort(partition_function, dim=1, descending=True)

                log_prob = (-energy_pos) - torch.logsumexp(partition_function, dim=1, keepdim=True)
                loss = (-log_prob).mean()

            # If the minimum idx is 0 then the ground truth configuration has the lowest energy
            acc = idx.eq(0).float().mean()
            accs.append(acc.item())
            losses.append(loss.item())

            node_pos, node_neg = node_pos.cpu().numpy(), node_neg.cpu().numpy()

            idx = idx.cpu().detach().numpy()
            sort_idx = sort_idx.cpu().detach().numpy()
            node_neg = np.reshape(node_neg, (-1, neg_sample, *node_neg.shape[1:]))

            for i in range(node_pos.shape[0]):
                gt_chi, chi_valid = gt_chis[i][0]

                if sort_idx[i, 0] == 0:
                    neg_chi, chi_valid = neg_chis[i][sort_idx[i, 1] - 1]

                else:
                    neg_chi, chi_valid = neg_chis[i][sort_idx[i, 0] - 1]

                neg_chi[neg_chi > 180] = neg_chi[neg_chi > 180] - 360
                score, max_dist = compute_rotamer_score_planar(gt_chi, neg_chi, chi_valid, res[i])
                rotamer_recovery.append(score.all())

                score = 0
                min_distance = float("inf")
                print_dist = None
                if not test:
                    for j, (neg_chi, chi_valid) in enumerate(neg_chis[i]):
                        temp_score, max_dist = compute_rotamer_score_planar(
                            gt_chi, neg_chi, chi_valid, res[i]
                        )
                        score = max(score, temp_score)

                        if max(max_dist) < min_distance:
                            min_distance = max(max_dist)
                            print_dist = max_dist
                else:
                    score = 1

                best_rotamer_recovery.append(score)


            if counter > 20 and (not test):
                # Return preliminary scores of rotamer recovery
                print(
                    "Mean cumulative accuracy of: ",
                    np.mean(accs),
                    np.std(accs) / np.sqrt(len(accs)),
                )
                print("Mean losses of: ", np.mean(losses), np.std(losses) / np.sqrt(len(losses)))
                print(
                    "Rotamer recovery ",
                    np.mean(rotamer_recovery),
                    np.std(rotamer_recovery) / np.sqrt(len(rotamer_recovery)),
                )
                print(
                    "Best Rotamer recovery ",
                    np.mean(best_rotamer_recovery),
                    np.std(best_rotamer_recovery) / np.sqrt(len(best_rotamer_recovery)),
                )
                break

    return np.mean(accs), np.mean(losses), np.mean(rotamer_recovery)


def main_single(gpu, FLAGS):
    if FLAGS.slurm:
        init_distributed_mode(FLAGS)

    os.environ["MASTER_ADDR"] = str(FLAGS.master_addr)
    os.environ["MASTER_PORT"] = str(FLAGS.port)

    rank_idx = FLAGS.node_rank * FLAGS.gpus + gpu
    world_size = FLAGS.nodes * FLAGS.gpus

    if rank_idx == 0:
        print("Values of args: ", FLAGS)

    if world_size > 1:
        if FLAGS.slurm:
            dist.init_process_group(
                backend="nccl", init_method="env://", world_size=world_size, rank=rank_idx
            )
        else:
            dist.init_process_group(
                backend="nccl",
                init_method="tcp://localhost:1492",
                world_size=world_size,
                rank=rank_idx,
            )

    train_dataset = MMCIFTransformer(
        FLAGS,
        split="train",
        rank_idx=rank_idx,
        world_size=world_size,
        uniform=FLAGS.uniform,
        weighted_gauss=FLAGS.weighted_gauss,
        gmm=FLAGS.gmm,
        chi_mean=FLAGS.chi_mean,
        mmcif_path=MMCIF_PATH,
    )
    valid_dataset = MMCIFTransformer(
        FLAGS,
        split="val",
        rank_idx=rank_idx,
        world_size=world_size,
        uniform=FLAGS.uniform,
        weighted_gauss=FLAGS.weighted_gauss,
        gmm=FLAGS.gmm,
        chi_mean=FLAGS.chi_mean,
        mmcif_path=MMCIF_PATH,
    )
    test_dataset = MMCIFTransformer(
        FLAGS,
        split="test",
        rank_idx=0,
        world_size=1,
        uniform=FLAGS.uniform,
        weighted_gauss=FLAGS.weighted_gauss,
        gmm=FLAGS.gmm,
        chi_mean=FLAGS.chi_mean,
        mmcif_path=MMCIF_PATH,
    )
    train_dataloader = DataLoader(
        train_dataset,
        num_workers=FLAGS.data_workers,
        collate_fn=collate_fn_transformer,
        batch_size=FLAGS.batch_size // FLAGS.multisample,
        shuffle=True,
        pin_memory=False,
        drop_last=True,
    )
    valid_dataloader = DataLoader(
        valid_dataset,
        num_workers=0,
        collate_fn=collate_fn_transformer_test,
        batch_size=FLAGS.batch_size // FLAGS.multisample,
        shuffle=True,
        pin_memory=False,
        drop_last=True,
    )
    test_dataloader = DataLoader(
        test_dataset,
        num_workers=0,
        collate_fn=collate_fn_transformer_test,
        batch_size=FLAGS.batch_size,
        shuffle=True,
        pin_memory=False,
        drop_last=True,
    )

    train_structures = train_dataset.files

    FLAGS_OLD = FLAGS

    logdir = osp.join(FLAGS.logdir, FLAGS.exp)

    if FLAGS.resume_iter != 0:
        model_path = osp.join(logdir, "model_{}".format(FLAGS.resume_iter))
        checkpoint = torch.load(model_path)
        try:
            FLAGS = checkpoint["FLAGS"]

            # Restore arguments to saved checkpoint values except for a select few
            FLAGS.resume_iter = FLAGS_OLD.resume_iter
            FLAGS.nodes = FLAGS_OLD.nodes
            FLAGS.gpus = FLAGS_OLD.gpus
            FLAGS.node_rank = FLAGS_OLD.node_rank
            FLAGS.master_addr = FLAGS_OLD.master_addr
            FLAGS.neg_sample = FLAGS_OLD.neg_sample
            FLAGS.train = FLAGS_OLD.train
            FLAGS.multisample = FLAGS_OLD.multisample
            FLAGS.steps = FLAGS_OLD.steps
            FLAGS.step_lr = FLAGS_OLD.step_lr
            FLAGS.batch_size = FLAGS_OLD.batch_size

            for key in dir(FLAGS):
                if "__" not in key:
                    FLAGS_OLD[key] = getattr(FLAGS, key)

            FLAGS = FLAGS_OLD
        except Exception as e:
            print(e)
            print("Didn't find keys in checkpoint'")

    if FLAGS.model == "transformer":
        import pdb
        pdb.set_trace()
        model = RotomerTransformerModel(FLAGS).train()
    elif FLAGS.model == "fc":
        model = RotomerFCModel(FLAGS).train()
    elif FLAGS.model == "s2s":
        model = RotomerSet2SetModel(FLAGS).train()
    elif FLAGS.model == "graph":
        model = RotomerGraphModel(FLAGS).train()

    if FLAGS.cuda:
        torch.cuda.set_device(gpu)
        model = model.cuda(gpu)

    optimizer = optim.Adam(model.parameters(), lr=FLAGS.start_lr, betas=(0.99, 0.999))

    if FLAGS.gpus > 1:
        sync_model(model)

    logger = TensorBoardOutputFormat(logdir)

    it = FLAGS.resume_iter

    if not osp.exists(logdir):
        os.makedirs(logdir)

    checkpoint = None
    if FLAGS.resume_iter != 0:
        model_path = osp.join(logdir, "model_{}".format(FLAGS.resume_iter))
        checkpoint = torch.load(model_path)
        try:
            optimizer.load_state_dict(checkpoint["optimizer_state_dict"])
            model.load_state_dict(checkpoint["model_state_dict"])
        except Exception as e:
            print("Transfer between distributed to non-distributed")

            model_state_dict = {
                k.replace("module.", ""): v for k, v in checkpoint["model_state_dict"].items()
            }
            model.load_state_dict(model_state_dict)

    pytorch_total_params = sum([p.numel() for p in model.parameters() if p.requires_grad])

    if rank_idx == 0:
        print("New Values of args: ", FLAGS)
        print("Number of parameters for models", pytorch_total_params)

    if FLAGS.train:
        train(
            train_dataloader,
            valid_dataloader,
            logger,
            model,
            optimizer,
            FLAGS,
            logdir,
            rank_idx,
            train_structures,
            checkpoint=checkpoint,
        )
    else:
        test(test_dataloader, model, FLAGS, logdir)


def main():
    # parse arguments
    parser = argparse.ArgumentParser()
    parser = add_args(parser)
    FLAGS = parser.parse_args()

    # convert to easy_dict; this is what is saved with model checkpoints and used in logic above
    keys = dir(FLAGS)
    flags_dict = EasyDict()
    for key in keys:
        if "__" not in key:
            flags_dict[key] = getattr(FLAGS, key)

    # postprocess arguments
    flags_dict.train = not flags_dict.no_train
    flags_dict.cuda = not flags_dict.no_cuda
    flags_dict.encoder_normalize_before = not flags_dict.no_encoder_normalize_before
    flags_dict.augment = not flags_dict.no_augment
    if not flags_dict.train:
        flags_dict.multisample = 1

    # Launch the job (optionally in a distributed manner)
    if flags_dict.gpus > 1:
        mp.spawn(main_single, nprocs=flags_dict.gpus, args=(flags_dict,))
    else:
        main_single(0, flags_dict)


if __name__ == "__main__":
    main()
